src/MetaMorpho.sol

```diff
@@ -327,9 +328,14 @@ contract MetaMorpho is ERC4626, ERC20Permit, Ownable2Step, Multicall, IMetaMorph
             if (!seen[i]) {
                 Id id = withdrawQueue[i];
 
-                if (MORPHO.supplyShares(id, address(this)) != 0 || config[id].cap != 0) {
-                    revert ErrorsLib.MissingMarket(id);
-                }
+                if (config[id].cap != 0) revert ErrorsLib.MissingMarket(id);
+
+                bool success;
+                try this.expectedSupplyAssets(_marketParams(id)) {
+                    success = true;
+                } catch {}
+
+                if (success && MORPHO.supplyShares(id, address(this)) != 0) revert ErrorsLib.MissingMarket(id);
```

Comment:
```suggestion
                try this.expectedSupplyAssets(_marketParams(id)) returns (uint256 supplyAssets) {
                  if (supplyAssets != 0) revert ErrorsLib.MissingMarket(id);
                } catch {}
```

Reply:
why? `supplyShares` is more accurate in this case

Reply:
But the following property holds: `supplyShares == 0 => supplyAssets == 0`
Reciprocally, `supplyAssets != 0 => supplyShares != 0`

So it's more optimized

Reply:
Right, my main concern is that we're doing twice the CALL to `MORPHO.supplyShares`

The minimum refactoring is:
```solidity
try this.expectedSupplyAssets(_marketParams(id)) {
  if (MORPHO.supplyShares(id, address(this)) != 0) revert ErrorsLib.MissingMarket(id);
} catch {}
```

Then from this point, we have the choice to:
1. do nothing
2. change `expectedSupplyAssets` to `expectedSupplyBalance() returns (uint256 shares, uint256 assets)` and use the return value `shares`

Reply:
The function `expectedSupplyBalance` that you mentioned does not exist though...

Reply:
We can create it, without using `MorphoBalancesLib.expectedSupplyAssets` directly!

Reply:
> We can create it, without using MorphoBalancesLib.expectedSupplyAssets directly!

I'll do this option on top

Reply:
We could also simply change the spec and allow removing a market from the withdraw queue as soon as the nb of shares on this market doesn't correspond to a single atom of the underlying asset

The reason we're limiting this from the start was to guarantee lenders the possibility to exit the vault at any time, given the liquidity on underlying markets. If the vault doesn't have a single atom of underlying asset supplied on a market, who cares about being able to withdraw from it?

With accrued interest, one may argue that this possibly non-zero nb of shares may grow to a non-zero supply of assets. To what extent? Let's imagine that interest grows 10 times the volume of the underlying market: the non-zero supply of shares would correspond to a maximum of 10 atoms of assets. It's just dust, so who cares about being able to withdraw this liquidity? The allocator can always reallocate it btw.

Reply:
Good point indeed, I think this is more that you still get expose to some extent to a market. 

But I agree that this is a ridiculously small amount and would simplify the code...

@Jean-Grimal do you have a though?

Reply:
What is a single atom ?

Reply:
This <img width="539" alt="Screenshot 2023-11-06 at 18 44 45" src="https://github.com/morpho-org/metamorpho/assets/44097430/c17b8cf5-4c61-4ed6-8493-9b68f87a92e5">

Joke aside, I think what @Rubilmax means is that if the shares are worth 0 tokens, then we don't care about those remaining shares and we can skip the market

Reply:
The indivisible unit of an asset (e.g. 1 wei for ETH)

Reply:
I agree with @Rubilmax, I don't think removing markets with less than 1 atom worth of supply is a problem.
And actually I think in over 99% of cases, markets with less than 1 atom worth of supply will be markets with 0 shares.

Reply:
I agree iff #172 is merged